### DISCUSSION TITLE
Documentation on how concurrency is handled is not clear

### DISCUSSION THREAD

**Discussion by darraghjones** [rabbitmq/rabbitmq-dotnet-client#1452] [General]
For example, the [docs](https://www.rabbitmq.com/dotnet-api-guide.html) do not mention [ConsumerDispatchConcurrency](https://rabbitmq.github.io/rabbitmq-dotnet-client/api/RabbitMQ.Client.ConnectionFactory.html#RabbitMQ_Client_ConnectionFactory_ConsumerDispatchConcurrency)

the [docs ](https://www.rabbitmq.com/dotnet-api-guide.html#consumer-callbacks-and-ordering)also say:

> callbacks are invoked concurrently

which does not appear to be the default behaviour (presumably because the default value for `ConsumerDispatchConcurrency` is 1?). Additionally they go on to say

> Consumer callbacks are invoked in tasks dispatched a [TaskScheduler](https://msdn.microsoft.com/en-us/library/dd997402%28v=vs.110%29.aspx).

which is pretty incomprehensible. Essentially it's very difficult for me (as an experienced dotnet software developer) to work out whether concurrency is handled by the library, or whether i should handle it in my application code.

**Comment by michaelklishin** [comment]
Concurrency is never "completely handled" by a library. This client won't prevent you from concurrency hazards when your consumers access shared resources or have any kind of inter-dependency.

What RabbitMQ client libraries do guarantee is

 * That deliveries on a **specific channel** are dispatched in the order of their arrival. Dispatch of deliveries on different channels can be interleaved, just like their frames are interleaved on the wire
 * That default concurrent dispatch level is 1, that is, the client won't run handlers for deliveries on a given channel concurrently and reduce the likelihood of uncalled for race conditions between delivery handlers

Whether #2 is true for the current generation of this client, I'm not sure but it helps reduce
the probability of surprising race conditions between delivery handlers. Of course, this library
cannot prevent your delivery handlers from offloading their work to thread pools or similar, potentially recreating a race condition in a different place.

**Comment by michaelklishin** [comment]
I have [tried to clarify it](https://github.com/rabbitmq/rabbitmq-website/commit/e618faccb9e6c7b4c810778a2de71e9c715ac1ef) somewhat.

**Comment by michaelklishin** [comment]
I don't know what specifically about the use of a `TaskScheduler` is "incomprehensible" but that's not really important, so I have removed the entire sentence.

The docs do not currently mention `IAsyncBasicConsumer` and `ConnectionFactory.DispatchConsumersAsync`. I assume both can be considered stable features in 7.0+. Their actual effects on consumer callback execution is something I'd need to investigate before attempting to document it.
